### ISSUE TITLE
DeltaSource load all files to Spark Driver during every batch. For big Table it's a huge issue.

### ISSUE THREAD

**Issue by GrigorievNick** [delta-io/delta#580]
This line of code uploads all files from delta-log to spark driver.
https://github.com/delta-io/delta/blob/819bcd5635dd47775d1c6adb51b3f40674d0ea81/src/main/scala/org/apache/spark/sql/delta/files/DeltaSourceSnapshot.scala#L86
1. This creates a very big overhead for spark driver memory and GC for big topics.
2. It's pretty strange to see collect inside a function that returns an iterator. Does any concern exist to do not use `toLocalIterator`?

**Comment by zsxwing** [comment]
Good catch. Yep, it makes sense to use `toLocalIterator`. Feel free to submit a PR to fix it.

**Comment by GrigorievNick** [comment]
thx @zsxwing, right now we use spark 2.4.7, and the master branch of the delta.io compatible only with spark 3.
So I have two questions. Please suggest me.
1. Can I do the fix as the patch to 0.6.1 tag and expect that it will be published with 0.6.2 delta.io version?
2. I will do a fix to spark 3 as well, but right now I don't have the environment to test this fix, so does it ok If I will do the fix without internal testing? Do the delta.io test handle streaming API or it's better to publish MR with enhancement after my team adopts spark 3?

**Comment by fvaleye** [comment]
Hello @GrigorievNick, I submitted a PR related to your issue (only with the latest version of delta-io, with Spark 3). I hope it helps!

**Comment by GrigorievNick** [comment]
Hi @fvaleye Thank you.
This does not affect my current project, because I use spark 2, and will proceed for the next few months.
So I just have my own fork of delta.io 0.6.1.
But this saves my time for contributing to things that I don't use tight now at work.
And also help to improve project quality in general.
So again, thank you. 
It was very helpful.

**Comment by foss-contributor** [comment]
@GrigorievNick, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
